File: Backend/app/services/noticia_service.py
import uuid
from datetime import datetime
from sqlalchemy.orm import Session
from app.models.models import Noticia


def listar(db: Session, solo_publicadas: bool = False) -> list[Noticia]:
    q = db.query(Noticia)
    if solo_publicadas:
        q = q.filter(Noticia.publicada == True)
    return q.order_by(Noticia.creado_en.desc()).all()


def obtener(db: Session, noticia_id: str) -> Noticia | None:
    return db.query(Noticia).filter(Noticia.id == noticia_id).first()


def crear(db: Session, titulo: str, contenido: str, categoria: str | None,
          imagen_url: str | None, autor_nombre: str | None, autor_cargo: str | None) -> Noticia:
    noticia = Noticia(
        id=uuid.uuid4(),
        titulo=titulo,
        contenido=contenido,
        categoria=categoria,
        imagen_url=imagen_url,
        autor_nombre=autor_nombre,
        autor_cargo=autor_cargo,
        publicada=False,
        likes=0,
        creado_en=datetime.utcnow(),
        actualizado_en=datetime.utcnow(),
    )
    db.add(noticia)
    db.commit()
    db.refresh(noticia)
    return noticia


def actualizar(db: Session, noticia_id: str, **kwargs) -> Noticia | None:
    noticia = obtener(db, noticia_id)
    if not noticia:
        return None
    for k, v in kwargs.items():
        if v is not None:
            setattr(noticia, k, v)
    noticia.actualizado_en = datetime.utcnow()
    db.commit()
    db.refresh(noticia)
    return noticia


def eliminar(db: Session, noticia_id: str) -> bool:
    noticia = obtener(db, noticia_id)
    if not noticia:
        return False
    db.delete(noticia)
    db.commit()
    return True


def toggle_like(db: Session, noticia_id: str, sumar: bool) -> Noticia | None:
    noticia = obtener(db, noticia_id)
    if not noticia:
        return None
    noticia.likes = max(0, noticia.likes + (1 if sumar else -1))
    db.commit()
    db.refresh(noticia)
    return noticia

File: Frontend/infodets-web/src/services/api/noticiaService.ts
import axiosInstance from '@/lib/axiosInstance'

export interface Noticia {
  id: string
  titulo: string
  contenido: string
  categoria?: string
  imagen_url?: string
  autor_nombre?: string
  autor_cargo?: string
  publicada: boolean
  likes: number
  creado_en: string
  actualizado_en: string
}

export const noticiaService = {
  listar: async (soloPublicadas = false): Promise<Noticia[]> => {
    const res = await axiosInstance.get<Noticia[]>('/noticias', { params: { solo_publicadas: soloPublicadas } })
    return res.data
  },

  crear: async (datos: FormData): Promise<Noticia> => {
    const res = await axiosInstance.post<Noticia>('/noticias', datos, {
      headers: { 'Content-Type': 'multipart/form-data' },
    })
    return res.data
  },

  actualizar: async (id: string, datos: FormData): Promise<Noticia> => {
    const res = await axiosInstance.put<Noticia>(`/noticias/${id}`, datos, {
      headers: { 'Content-Type': 'multipart/form-data' },
    })
    return res.data
  },

  eliminar: async (id: string): Promise<void> => {
    await axiosInstance.delete(`/noticias/${id}`)
  },

  like: async (id: string, sumar: boolean): Promise<Noticia> => {
    const res = await axiosInstance.post<Noticia>(`/noticias/${id}/like`, null, { params: { sumar } })
    return res.data
  },
}
